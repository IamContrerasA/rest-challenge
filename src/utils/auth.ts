import jwt from 'jsonwebtoken'
import { PrismaClient, User } from '@prisma/client'
import { Request, Response, NextFunction } from 'express'
import bcrypt from 'bcrypt'

const prisma = new PrismaClient()
const JWT_SECRET = 'JWT_SECRET'

export const newToken = (user: User | null) => {
  return jwt.sign({ id: user?.id }, JWT_SECRET, {
    expiresIn: '100d',
  })
}

export const verifyToken = (token: string) =>
  new Promise((resolve, reject) => {
    jwt.verify(token, JWT_SECRET, (err, payload) => {
      if (err) return reject(err)
      resolve(payload)
    })
  })

export const signup = async (req: Request, res: Response): Promise<void> => {
  if (!req.body.email || !req.body.password) {
    res.status(400).send({ message: 'need email and password' })
  }

  try {
    let user: User
    bcrypt.hash(req.body.password, 8, async (err, hash) => {
      if (err) {
        throw new Error(`There is an error on hash password ${err}`)
      }
      user = await prisma.user.create({
        data: {
          email: req.body.email,
          password: hash,
        },
      })
    })

    const token = newToken(user!)
    res.status(201).send({ token })
  } catch (e) {
    return res.status(500).end()
  }
}

//checkpassword
function checkPassword(userPassord: string, password: string) {
  return new Promise((resolve, reject) => {
    bcrypt.compare(password, userPassord, (err, same) => {
      if (err) {
        console.log(err)
        return reject(err)
      }
      resolve(same)
    })
  })
}

export async function signin(req: Request, res: Response): Promise<void> {
  if (!req.body.email || !req.body.password) {
    res.status(400).send({ message: 'need email and password' })
  }

  const invalid = 'Invalid email and passoword combination'

  try {
    const user = await prisma.user.findUnique({
      where: { email: req.body.email },
    })

    if (!user) {
      res.status(401).send({ message: `${invalid}, no user found` })
    }

    const match = await checkPassword(user!.password, req.body.password)

    if (!match) {
      res.status(401).send({ message: `${invalid}, password dont match` })
    }

    const token = newToken(user)
    res.status(201).send({ token })
  } catch (e) {
    console.error(e)
    res.status(500).end()
  }
}

export const protect = async (
  req: Request,
  res: Response,
  next: NextFunction,
): Promise<void> => {
  const bearer = req.headers.authorization

  if (!bearer || !bearer.startsWith('Bearer ')) {
    return res.status(401).end()
  }

  const token = bearer.split('Bearer ')[1].trim()
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  let payload: any
  try {
    payload = await verifyToken(token)
  } catch (e) {
    return res.status(401).end()
  }

  const user = await prisma.user.findUnique({
    where: { id: payload.id },
    select: {
      id: true,
      firstname: true,
      surname: true,
      email: true,
      password: false,
    },
  })
  if (!user) {
    return res.status(401).end()
  }

  req.body.user = user
  next()
}
