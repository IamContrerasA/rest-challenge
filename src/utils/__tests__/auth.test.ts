/* eslint-disable @typescript-eslint/no-explicit-any */
import { prisma, protect, signin, signup } from '../../utils/auth'

let token: string

afterAll(async () => {
  await prisma.$executeRaw(
    `TRUNCATE TABLE public."Comment" RESTART IDENTITY CASCADE;`,
  )
  await prisma.$executeRaw(
    `TRUNCATE TABLE public."Post" RESTART IDENTITY CASCADE;`,
  )
  await prisma.$executeRaw(
    `TRUNCATE TABLE public."User" RESTART IDENTITY CASCADE;`,
  )
  await prisma.$disconnect()
})

describe('auth happy way', () => {
  test('signup with correct values is valid', async () => {
    const req: any = {
      body: {
        email: 'myemail16@email.com',
        password: 'password16',
      },
    }

    const res: any = {
      status: function (statusNumber: number) {
        expect(statusNumber).toBe(201)
        return this
      },
      send: function (sendResult: any) {
        expect(sendResult.token.length).toBe(137)
        return this
      },
    }
    await signup(req, res)
  })

  test('signin with correct values is valid', async () => {
    await new Promise((r) => setTimeout(r, 500))
    const req: any = {
      body: {
        email: 'myemail16@email.com',
        password: 'password16',
      },
    }

    const res: any = {
      status: function (statusNumber: number) {
        expect(statusNumber).toBe(201)
        return this
      },
      send: function (sendResult: any) {
        expect(sendResult.token.length).toBe(137)
        token = sendResult.token
        return this
      },
    }
    await signin(req, res)
  })

  test('protect with correct values is valid', async () => {
    const req: any = {
      body: {},
      headers: { authorization: `Bearer ${token}` },
    }

    const res: any = {
      status: function (statusNumber: number) {
        expect(statusNumber).toBe(200)
        return this
      },
      send: function (sendResult: any) {
        expect(sendResult.user.id).toBe(1)
        token = sendResult.token
        return this
      },
    }
    await protect(req, res, jest.fn())
  })
})
