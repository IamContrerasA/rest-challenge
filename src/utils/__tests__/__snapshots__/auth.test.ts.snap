// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auth error way signin with incorrect values is invalid 1`] = `"error on password hash"`;

exports[`auth error way signin without email or password  is invalid 1`] = `"need email and password"`;

exports[`auth error way signout with incorrect values is valid 1`] = `"email already exists"`;

exports[`auth error way signup with incorrect values is invalid 1`] = `"need email and password"`;

exports[`auth error way verify email with incorrect values is invalid 1`] = `"User not found"`;

exports[`auth happy way protect without signin correct values is valid 1`] = `"Please signin first"`;

exports[`auth happy way signout with correct values is valid 1`] = `"user logout"`;

exports[`auth happy way verify email with correct values is valid 1`] = `"Email verified!"`;
